src/animals/animals.service.ts

```diff
@@ -1,15 +1,162 @@
 
 @Injectable()
 export class AnimalsService {
   constructor(
     @InjectModel(Animal.name) private animalModel: Model<AnimalDocument>,
+    private readonly redisCacheManager: RedisCacheService,
   ) {}
 
-  async findAll() {
-    return this.animalModel.find();
+  async findAll(): Promise<Animal[]> {
+    try {
+      return await this.animalModel.find();
+    } catch (err) {
+      // console.error(err);
+      throw new InternalServerErrorException();
+    }
+  }
+
+  async findOne(id: string): Promise<Animal> {
+    try {
+      const animal: AnimalDocument = await this.animalModel.findById({
+        _id: new Types.ObjectId(id),
+      });
+
+      if (!animal) {
+        throw new NotFoundException();
+      }
+      return animal;
+    } catch (err) {
+      // console.error(err);
+      if (err instanceof NotFoundException) {
+        throw new NotFoundException();
+      }
+      throw new InternalServerErrorException();
+    }
+  }
+
+  async update(id: string, animalData: UpdateAnimalDto): Promise<Animal> {
+    try {
+      const animal: AnimalDocument = await this.animalModel.findByIdAndUpdate(
+        { _id: new Types.ObjectId(id) },
+        animalData,
+        {
+          new: true,
+        },
+      );
+
+      if (!animal) {
+        throw new NotFoundException();
+      }
+
+      await this.redisCacheManager.clearCache();
+      return animal;
+    } catch (err) {
+      console.error(err);
+      if (err instanceof NotFoundException) {
+        throw new NotFoundException();
+      }
+      throw new InternalServerErrorException();
+    }
+  }
+
+  async create(animalData: AnimalDto): Promise<Animal> {
+    try {
+      const animal: AnimalDocument = await this.animalModel.findOne({
+        animalName: { $eq: animalData.animalName },
+        animalType: { $eq: animalData.type },
+      });
+
+      if (animal) {
+        throw new BadRequestException('Animal already exist in database');
+      }
+
+      const newAnimal: AnimalDocument = await this.animalModel.create({
+        createdAt: new Date(),
+        ...animalData,
+      });
+
+      await this.redisCacheManager.clearCache();
+
+      return newAnimal;
+    } catch (err) {
+      // console.error(err);
+      if (err instanceof BadRequestException) {
+        throw new BadRequestException(err.message);
+      }
+      throw new InternalServerErrorException();
+    }
+  }
+
+  async addAnimals(animalsList: AnimalDto[]): Promise<Animal[]> {
+    try {
+      const allAnimals: Animal[] = await this.findAll();
+      const alreadyExistAnimals: AnimalDto[] = animalsList.filter((animal) =>
+        allAnimals.some(
+          ({ animalName, type }) =>
+            animal.animalName === animalName && animal.type === type,
+        ),
+      );
+      if (alreadyExistAnimals.length !== 0) {
+        throw new BadRequestException('Animals already exist in database');
+      }
+
+      await this.redisCacheManager.clearCache();
```

Comment:
czemu tak? nie chcemy usuwac tych ktore juz sa

Reply:
Ja podszedłem do tego tak, że jak zachodzi jakaś zmiana w bazie danych to resetuje to co jest w Cache, aby przy nowym zapytaniu pobrać aktualną bazę, a nie to co jest w Cache (nieaktualne)

Reply:
Poprawiłem i powinno być już dobrze :)

Reply:
gdzie to masz wrzucone bo nie widze zmian

Reply:
w RedisCache.service.ts mam zaimplementowaną metodę clearCache, która czyści zapytanie findAll i przy update zapytanie konkretnego zapytania findOne o podanym id

Reply:
Podsumowując metoda clearCache z RedisCacheService, usuwa za każdym razem klucz findAll, przy zapytaniach POST i PATCH i dodatkowo usuwa klucz pojedynczego zapytania findOne dla danego id przy zapytaniu PATCH